src/leaderboard/leaderboard_resource.rs

```diff
@@ -0,0 +1,128 @@
+        request.send(self).await
+    }
+
+    /// Fetch elements from a leaderboard by rank.
+    pub async fn fetch_by_rank(
+        &self,
+        rank_range: impl Into<RankRange>,
+    ) -> MomentoResult<FetchResponse> {
+        let request = FetchByRankRequest::new(rank_range);
+        request.send(self).await
+    }
+
+    /// Get elements from a leaderboard by score.
+    pub async fn fetch_by_score(
+        &self,
+        score_range: impl Into<ScoreRange>,
+    ) -> MomentoResult<FetchResponse> {
+        let request = FetchByScoreRequest::new(score_range);
+        request.send(self).await
+    }
+
+    /// Get the length of a leaderboard.
+    pub async fn length(&self) -> MomentoResult<LengthResponse> {
+        let request = LengthRequest::new();
+        request.send(self).await
+    }
+
+    /// Get rank of elements from a leaderboard using their element ids.
+    pub async fn get_rank<T: IntoIds>(&self, ids: T) -> MomentoResult<GetRankResponse> {
+        let request = GetRankRequest::new(ids);
+        request.send(self).await
+    }
+
+    /// Remove elements from a leaderboard using their element ids.
+    pub async fn remove_elements<T: IntoIds>(
+        &self,
+        ids: T,
+    ) -> MomentoResult<RemoveElementsResponse> {
+        let request = RemoveElementsRequest::new(ids);
+        request.send(self).await
+    }
+
+    /// Upsert (update/insert) elements into a leaderboard.
+    pub async fn upsert<E: IntoElements>(&self, elements: E) -> MomentoResult<UpsertResponse> {
+        let request = UpsertRequest::new(elements);
+        request.send(self).await
+    }
+
+    /* helper fns */
+    pub(crate) fn new(
+        data_clients: Vec<SLbClient<InterceptedService<Channel, HeaderInterceptor>>>,
+        deadline: Duration,
+        cache_name: impl Into<String>,
+        leaderboard_name: impl Into<String>,
+    ) -> Self {
+        Self {
+            data_clients,
+            deadline,
+            cache_name: cache_name.into(),
+            leaderboard_name: leaderboard_name.into(),
+        }
+    }
+
+    pub(crate) fn next_data_client(
+        &self,
+    ) -> SLbClient<InterceptedService<Channel, HeaderInterceptor>> {
+        let next_index =
+            NEXT_DATA_CLIENT_INDEX.fetch_add(1, Ordering::Relaxed) % self.data_clients.len();
+        self.data_clients[next_index].clone()
+    }
+
+    pub(crate) fn deadline(&self) -> Duration {
+        self.deadline
+    }
+
+    pub(crate) fn cache_name(&self) -> &String {
+        &self.cache_name
+    }
+
+    pub(crate) fn leaderboard_name(&self) -> &String {
+        &self.leaderboard_name
+    }
+
+    /// Lower-level API to send any type of MomentoRequest to the server. This is used for cases when
+    /// you want to set optional fields on a request that are not supported by the short-hand API for
+    /// that request type.
+    pub async fn send_request<R: MomentoRequest>(&self, request: R) -> MomentoResult<R::Response> {
```

Comment:
This trait name is rough, and overloaded with other MomentoRequest. It's going to be hard to read in user code. Consider using something non-overlapping and clear like `LeaderboardRequest` instead.

Reply:
Good idea. I updated this in [87dc81a](https://github.com/momentohq/client-sdk-rust/pull/438/commits/87dc81a20d361e0df05e9314f12c55bacb8ac939). We still have a type def for `MomentoResult<T> = Result<T, MomentoError>` which is shared across the clients.